src/api/procedures/setAssetStats.ts

```diff
@@ -1,29 +1,98 @@
-import { PolymeshPrimitivesStatisticsStatType } from '@polkadot/types/lookup';
+import {
+  PolymeshPrimitivesStatisticsStatType,
+  PolymeshPrimitivesStatisticsStatUpdate,
+} from '@polkadot/types/lookup';
+import { BTreeSet } from '@polkadot/types-codec';
 
-import { Procedure } from '~/internal';
-import { FungibleAsset, SetTransferRestrictionStatParams, StatType, TxTags } from '~/types';
-import { BatchTransactionSpec, ProcedureAuthorization, TxWithArgs } from '~/types/internal';
+import { Context, PolymeshError, Procedure } from '~/internal';
+import {
+  AddBalanceStatParams,
+  AddClaimBalanceStatParams,
+  AddClaimCountStatParams,
+  AddCountStatParams,
+  ErrorCode,
+  FungibleAsset,
+  SetTransferRestrictionStatParams,
+  StatType,
+  TxTag,
+  TxTags,
+} from '~/types';
+import { BatchTransactionSpec, ProcedureAuthorization } from '~/types/internal';
 import {
   assetToMeshAssetId,
+  balanceStatInputToStatUpdates,
+  claimBalanceStatInputToStatUpdates,
   claimCountStatInputToStatUpdates,
   claimIssuerToMeshClaimIssuer,
   countStatInputToStatUpdates,
+  inputStatTypeToMeshStatType,
   statisticsOpTypeToStatType,
   statisticStatTypesToBtreeStatType,
   statTypeToStatOpType,
 } from '~/utils/conversion';
-import { checkTxType } from '~/utils/internal';
+import { areBtreeSetsEqual, checkTxType } from '~/utils/internal';
 
 /**
  * @hidden
  */
 export type SetAssetStatParams = { asset: FungibleAsset } & SetTransferRestrictionStatParams;
 
+/**
+ * @hidden
+ */
+export interface SetAssetStatsStorage {
+  currentStats: BTreeSet<PolymeshPrimitivesStatisticsStatType>;
+}
+
+/**
+ * @hidden
+ */
+function getStatUpdates(
+  stat:
+    | AddCountStatParams
+    | AddBalanceStatParams
+    | AddClaimCountStatParams
+    | AddClaimBalanceStatParams,
+  context: Context
+): BTreeSet<PolymeshPrimitivesStatisticsStatUpdate> | undefined {
+  const { type } = stat;
+
+  if (type === StatType.Count && stat.count !== undefined) {
+    return countStatInputToStatUpdates(stat, context);
+  }
+
+  if (type === StatType.ScopedCount && stat.value !== undefined) {
+    return claimCountStatInputToStatUpdates(stat, context);
+  }
+
+  if (type === StatType.Balance && stat.balance !== undefined) {
+    return balanceStatInputToStatUpdates(stat, context);
+  }
+
+  if (type === StatType.ScopedBalance && stat.value !== undefined) {
+    return claimBalanceStatInputToStatUpdates(stat, context);
+  }
+
+  if (
+    type !== StatType.Count &&
+    type !== StatType.ScopedCount &&
+    type !== StatType.Balance &&
+    type !== StatType.ScopedBalance
+  ) {
+    throw new PolymeshError({
+      code: ErrorCode.UnexpectedError,
+      message: `Unsupported stat type: ${type}. Please report this to the Polymesh team`,
+    });
+  }
```

Comment:
Is this check really needed? Feels like something TS will guard against.

Wouldn't this only be triggered if a dev uses JS or an `any` assertion or the like? If type is coming from the user that's on them, so I don't think we'd be able to do much if this error was ever actually seen.